fontbe/src/features.rs

```diff
@@ -219,14 +274,171 @@ impl<'a> FeatureWriter<'a> {
+        let lookups = vec![builder.add_lookup(LookupFlag::empty(), None, vec![ppos_subtables])];
+        builder.add_to_default_language_systems(Tag::new(b"kern"), &lookups)
+    }
+
+    /// Generate mark to base and mark to mark features
+    ///
+    /// Based on notes from f2f at W3C TPAC Spain and inspection of fea written by fontmake.
+    ///
+    /// Emit one lookup per mark class, it's simpler and may be more compact. See discussions in:
+    /// * <https://github.com/googlefonts/ufo2ft/issues/762>
+    /// * <https://github.com/googlefonts/ufo2ft/issues/591>
+    /// * <https://github.com/googlefonts/ufo2ft/issues/563>
+    //TODO: could we generate as a separate task, and then just add here.
+    fn add_marks(&self, builder: &mut FeatureBuilder) -> Result<(), Error> {
+        let mut anchors = HashMap::new();
+        for (work_id, arc_glyph_anchors) in self.raw_anchors {
+            let glyph_anchors: &GlyphAnchors = arc_glyph_anchors;
+            let FeWorkId::Anchor(glyph_name) = work_id else {
+                return Err(Error::ExpectedAnchor(work_id.clone()));
+            };
+            anchors.insert(glyph_name.clone(), glyph_anchors);
         }
+
+        let mut groups: HashMap<GlyphName, MarkGroup> = Default::default();
+        for (glyph_name, glyph_anchors) in anchors.iter() {
+            // We assume the anchor list to be small
+            // considering only glyphs with anchors,
+            //  - glyphs with *only* base anchors are bases
+            //  - glyphs with *any* mark anchor are marks
+
+            let mut base = true; // TODO: only a base if user rules allow it
+            for anchor in glyph_anchors.anchors.iter() {
+                let anchor_type = anchor.name.anchor_type();
+                if anchor_type == AnchorType::Mark {
+                    base = false;
+
+                    // TODO: only if user rules allow us to be a mark
+                    groups
+                        .entry(anchor_type.group_name(anchor.name.clone()))
+                        .or_default()
+                        .marks
+                        .push(((glyph_name).clone(), anchor));
+                }
+            }
+
+            if base {
+                for anchor in glyph_anchors.anchors.iter() {
+                    let anchor_type = anchor.name.anchor_type();
+                    groups
+                        .entry(anchor_type.group_name(anchor.name.clone()))
+                        .or_default()
+                        .bases
+                        .push(((glyph_name).clone(), anchor));
+                }
+            }
+        }
+
+        // Build the actual mark base and mark mark constructs using fea-rs builders
+
+        let mut mark_base_lookups = Vec::new();
+        let mut mark_mark_lookups = Vec::new();
+
+        for (group_name, group) in groups.iter() {
+            let mark_class_name: SmolStr = format!("MC_{group_name}").into();
```

Comment:
I think you should be able to just use the group name directly now.

Reply:
Done. Making it a SmolStr still ends up needing an into.